File: J9Ch/src/J_3_Binary_Tree_Divide_Conquer/Required_10/_596_Minimum_Subtree_Easy.java
package J_3_Binary_Tree_Divide_Conquer.Required_10;
import java.util.*;import lib.TreeNode;import lib.AssortedMethods;import org.junit.Test;
import lib.TreeNode;
import lib.AssortedMethods;
import org.junit.Test;
/** 596. Minimum Subtree
 * Easy

 * Created by tianhuizhu on 6/27/17.
 */
public class _596_Minimum_Subtree_Easy {

    // version 1 : traverse + divide conquer
    private TreeNode subtree = null;
    private int subtreeSum = Integer.MAX_VALUE;
    /**
     * @param root the root of binary tree
     * @return the root of the minimum subtree
     */
    public TreeNode findSubtree(TreeNode root) {
        helper(root);
        return subtree;
    }

    private int helper(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int sum = helper(root.left) + helper(root.right) + root.val;
        if (sum <= subtreeSum) {
            subtreeSum = sum;
            subtree = root;
        }
        return sum;
    }


    // version 2: Pure divide conquer
    class ResultType {
        public TreeNode minSubtree;
        public int sum, minSum;
        public ResultType(TreeNode minSubtree, int minSum, int sum) {
            this.minSubtree = minSubtree;
            this.minSum = minSum;
            this.sum = sum;
        }
    }

    /**
     * @param root the root of binary tree
     * @return the root of the minimum subtree
     */
    public TreeNode findSubtree_2(TreeNode root) {
        ResultType result = helper_2(root);
        return result.minSubtree;
    }

    public ResultType helper_2(TreeNode node) {
        if (node == null) {
            return new ResultType(null, Integer.MAX_VALUE, 0);
        }

        ResultType leftResult = helper_2(node.left);
        ResultType rightResult = helper_2(node.right);

        ResultType result = new ResultType(
                node,
                leftResult.sum + rightResult.sum + node.val,
                leftResult.sum + rightResult.sum + node.val
        );

        if (leftResult.minSum <= result.minSum) {
            result.minSum = leftResult.minSum;
            result.minSubtree = leftResult.minSubtree;
        }

        if (rightResult.minSum <= result.minSum) {
            result.minSum = rightResult.minSum;
            result.minSubtree = rightResult.minSubtree;
        }

        return result;
    }

}

File: J9Ch/src/J_3_Binary_Tree_Divide_Conquer/Required_10/python_cpp/_596_Minimum_Subtree_Easy.cpp
/** 596. Minimum Subtree
 * Easy*/


class Solution {
public:
    /**
     * @param root the root of binary tree
     * @return the minimun weight node
     */
    TreeNode* findSubtree(TreeNode* root) {
        // Write your code here
        helper(root);
        return result;
    }

    int helper(TreeNode* root) {
        if (root == NULL) {
            return 0;
        }
        int left_weight = helper(root->left);
        int right_weight = helper(root->right);

        if (left_weight + right_weight + root->val <= minumun_weight) {
            minumun_weight = left_weight + right_weight + root->val;
            result = root;
        }
        return left_weight + right_weight + root->val;
    }

private:
    TreeNode* result = nullptr;
    int minumun_weight = INT_MAX;
};